### DISCUSSION TITLE
neorv32 and a generic FAT file system called FatFs.

### DISCUSSION THREAD

**Discussion by emb4fun** [stnolting/neorv32#606] [Show and tell]
Hello,

and another small component which I would like to publish here. This time there is no website as a description but only a small readme.

FatFs was used here as the file system, and the SD card is controlled in SPI 1-bit mode. There is a separate SPI master for this, which can be switched between 8 and 32 bits. This is part of the Qsys system here, but should also work as a Wishbone device with minor modifications. FatFs comes without a low level disk I/O module.

The following two files are now available for this purpose:

- fatfs_diskio_spi.c
- fatfs_neorv32_de0n_spi.c

A microSD SanDisk Ultra 16GB was used as the SD card here with a SD card holder which was connected by jumper wire. If everything works, the output on the terminal should look like this:

```
<<< FatFs filesystem benchmark >>>

Benchmark started ...

15541728 KiB total drive space.
15537488 KiB available.

Simple write/read test... OK

Performance write/read:
Write: 0.93 MB/s
Read : 1.40 MB/s

Benchmark end.
```
The project is available here on GitHub:
https://github.com/emb4fun/neorv32-de0n-fatfs

Best regards,
Michael

**Comment by stnolting** [comment]
Cool project and interesting benchmark results! 1.4MB/s for reading is pretty good for a plain 1-bit SPI interface! :+1: What's the block size the benchmark / file system is using? 512 bytes?

**Comment by emb4fun** [comment]
The block size / buffer size was 4096 bytes.

**Reply by stnolting** [reply]
That's quite large (making the performance results even more interesting)! 😉
